services/auth/service/oauth_service.go

```diff
@@ -146,3 +163,49 @@ func GetLastName(oauth_token string, provider string) (string, error) {
 		return "", errors.New("Invalid provider")
 	}
 }
+
+/*
+	A helper function that takes a URL pointer and a map of query params->values, and modifies the URL's
+	RawQuery property with the supplied query params.
+*/
+func ConstructURLQuery(u *url.URL, params map[string]string) {
+	q := u.Query()
+
+	for param, value := range params {
+		q.Set(param, value)
+	}
+
+	u.RawQuery = q.Encode()
+}
+
+/*
+	This function takes in the ingredients to a URL and outputs a string of them all together.
+	It also checks for the appearance of "#" anywhere in the query params and throws an error if it is there.
+    queryParams is an optional param. nil can be passed in if the url needs no query params.
+*/
+func ConstructSafeURL(scheme string, host string, path string, queryParams map[string]string) (string, error) {
+	url := url.URL{
+		Scheme: scheme,
+		Host:   host,
+		Path:   path,
+	}
+
+	// Per the OAuth 2.0 RFC 6749, we need to disallow the `#` fragment character in the URL
+	valid := true
+	if queryParams != nil {
+		for _, val := range queryParams {
+			if strings.Contains(val, "#") {
+				valid = false
+				break
```

Comment:
Why not just early return here instead of having a bool for control flow?

Reply:
Sorry, that was an artefact of my earlier loop idea, and I forgot to remove it!